R/lang-call.R

```diff
@@ -1,232 +1,121 @@
   }
-  formals_pool <- setdiff(formals_nms, actuals_nms)
-
-  is_matched <- actuals_nms %in% formals_nms
-  actuals_pat <- actuals_nms
-  actuals_pat[is_matched | is_empty] <- NA
 
-  matched <- list()
-  for (formal in formals_pool) {
-    matched_pos <- which(map_lgl(actuals_pat, arg_match_partial, formal))
-
-    if (length(matched_pos) > 1) {
-      stop(call. = FALSE,
-        "formal argument `", formal,
-        "` matched by multiple actual arguments")
-    }
-    if (length(matched_pos) && matched_pos %in% matched) {
-      stop(call. = FALSE,
-        "actual argument `", actuals_pat[matched_pos],
-        "` matches multiple formal arguments")
+  if (!is_null(name)) {
+    # Wrap language objects in a list
+    if (!is_vector(name)) {
+      name <- list(name)
     }
 
-    matched <- append(matched, matched_pos)
-    actuals_nms[matched_pos] <- formal
-  }
-
-  if (is.na(dots_pos)) {
-    is_unused <- !actuals_nms %in% c(formals_nms, "")
-    is_unused <- is_unused & !map_lgl(actuals_nms, is_dot_nm)
-    if (any(is_unused)) {
-      stop(call. = FALSE,
-        "unused arguments: ",
-        paste0(actuals_nms[is_unused], collapse = ", "))
+    unmatched <- TRUE
+    for (elt in name) {
+      if (identical(x[[1]], as_symbol(elt))) {
+        unmatched <- FALSE
+        break
+      }
     }
-    if (length(actuals_nms) > length(formals_nms)) {
-      stop("unused arguments", call. = FALSE)
+
+    if (unmatched) {
+      return(FALSE)
     }
   }
 
-  names(call) <- c("", actuals_nms)
-  call
-}
-
-call_inline_dots <- function(call, caller_env, enum_dots) {
-  d <- lsp_walk_nonnull(call, function(arg) {
-    if (identical(cadr(arg), quote(...))) arg
-  })
-  if (is.null(d)) {
-    return(call)
+  if (!is_null(n) && !has_length(x, n + 1L)) {
+    return(FALSE)
   }
-  if (is.null(caller_env)) {
-    stop(call. = FALSE, "`caller_env` must be supplied to match dots")
-  }
-
-  dots <- frame_dots_lsp(caller_env)
-  dots <- dots_enumerate_args(dots)
-
-  # Attach remaining args to expanded dots
-  remaining_args <- cddr(d)
-  lsp_walk_nonnull(dots, function(arg) {
-    if (is.null(cdr(arg))) set_cdr(arg, remaining_args)
-  })
 
-  # Replace dots symbol with actual dots and remaining args
-  set_cdr(d, dots)
-
-  call
+  TRUE
 }
-
-call_match <- function(call, fn, enum_dots, add_missings) {
-  args <- call[-1]
-  args_nms <- names2(args)
-  fmls_nms <- names2(fn_fmls(fn))
-
-  dots_i <- which(fmls_nms == "...")
-  if (length(dots_i)) {
-    args_nms <- call_match_dotted(args_nms, fmls_nms, dots_i, enum_dots)
-  } else {
-    args_nms <- call_match_args(args_nms, fmls_nms)
-  }
-  names(call) <- c("", args_nms)
-
-  if (add_missings) {
-    missing_nms <- setdiff(fmls_nms, c(args_nms, "..."))
-    missing_args <- rep(list(arg_missing()), length(missing_nms))
-    missing_args <- as.pairlist(set_names(missing_args, missing_nms))
-    call <- lsp_append(call, missing_args)
-  }
-
-  call
+#' @rdname is_lang
+#' @export
+is_language <- is_lang
```

Comment:
Maybe not, I thought it'd be nice to have a predicate that corresponds to the sexp type.